src/providers/simulatedUser.ts

```diff
@@ -58,40 +75,63 @@ export class SimulatedUser implements ApiProvider {
   private async sendMessageToAgent(
     messages: Message[],
     targetProvider: ApiProvider,
-    prompt: string,
-    context?: CallApiContextParams,
-  ): Promise<Message[]> {
-    const response = await targetProvider.callApi(
-      JSON.stringify([{ role: 'system', content: prompt }, ...messages]),
-      context,
-    );
+    context: CallApiContextParams,
+  ): Promise<ProviderResponse> {
+    const targetPrompt = this.stateful
+      ? messages[messages.length - 1].content
+      : JSON.stringify(messages);
+
+    logger.debug(`[SimulatedUser] Sending message to target provider: ${targetPrompt}`);
+
+    const response = await targetProvider.callApi(targetPrompt, context);
+
+    if (response.sessionId) {
+      context = context ?? { vars: {}, prompt: { raw: '', label: 'target' } };
+      context.vars.sessionId = response.sessionId;
+    }
+
     if (targetProvider.delay) {
       logger.debug(`[SimulatedUser] Sleeping for ${targetProvider.delay}ms`);
       await sleep(targetProvider.delay);
     }
-    logger.debug(`Agent: ${response.output}`);
-    return [...messages, { role: 'assistant', content: String(response.output || '') }];
+
+    logger.debug(`[SimulatedUser] Agent: ${response.output}`);
+    return response;
   }
 
   async callApi(
+    // NOTE: `prompt` is not used in this provider; `vars.instructions` is used instead.
     prompt: string,
     context?: CallApiContextParams,
     _callApiOptions?: CallApiOptionsParams,
   ): Promise<ProviderResponse> {
     invariant(context?.originalProvider, 'Expected originalProvider to be set');
 
     const instructions = getNunjucksEngine().renderString(this.rawInstructions, context?.vars);
-    const userProvider = new PromptfooSimulatedUserProvider({
-      instructions,
-    });
+    const userProvider = new PromptfooSimulatedUserProvider({ instructions }, this.taskId);
 
-    logger.debug(`Formatted user instructions: ${instructions}`);
-    let messages: Message[] = [];
+    logger.debug(`[SimulatedUser] Formatted user instructions: ${instructions}`);
+    const messages: Message[] = [];
     const maxTurns = this.maxTurns;
-    let numRequests = 0;
+
+    const totalTokenUsage = {
+      total: 0,
+      prompt: 0,
+      completion: 0,
+      numRequests: 0,
+      cached: 0,
+    };
+
+    let agentResponse: ProviderResponse | undefined;
+
     for (let i = 0; i < maxTurns; i++) {
+      logger.debug(`[SimulatedUser] Turn ${i + 1} of ${maxTurns}`);
```

Comment:
recommend cleaning up some of these logs

Reply:
I can do that but actually found these logs very helpful when trying to grok the state during a given turn.